hamt/util.go

```diff
@@ -0,0 +1,129 @@
+package hamt
+
+// adapted from https://github.com/ipfs/go-unixfs/blob/master/hamt/util.go
+
+import (
+	"fmt"
+
+	"math/bits"
+
+	"github.com/Stebalien/go-bitfield"
+	"github.com/ipfs/go-unixfsnode/data"
+	dagpb "github.com/ipld/go-codec-dagpb"
+	"github.com/spaolacci/murmur3"
+)
+
+// hashBits is a helper that allows the reading of the 'next n bits' as an integer.
+type hashBits struct {
+	b        []byte
+	consumed int
+}
+
+func mkmask(n int) byte {
+	return (1 << uint(n)) - 1
+}
+
+// Next returns the next 'i' bits of the hashBits value as an integer, or an
+// error if there aren't enough bits.
+func (hb *hashBits) Next(i int) (int, error) {
+	if hb.consumed+i > len(hb.b)*8 {
+		return 0, ErrHAMTTooDeep
+	}
+	return hb.next(i), nil
+}
+
+func (hb *hashBits) next(i int) int {
+	curbi := hb.consumed / 8
+	leftb := 8 - (hb.consumed % 8)
+
+	curb := hb.b[curbi]
+	if i == leftb {
+		out := int(mkmask(i) & curb)
+		hb.consumed += i
+		return out
+	} else if i < leftb {
```

Comment:
nit: unnecessary else

Reply:
good point -- copied code but you are totally right.